### DISCUSSION TITLE
Why should avoid launching synchronous subtasks?

### DISCUSSION THREAD

**Discussion by 954-Ivory** [celery/celery#7658] [Q&A]
[Having a task wait for the result of another task is really inefficient, and may even cause a deadlock if the worker pool is exhausted.](https://docs.celeryq.dev/en/stable/userguide/tasks.html#avoid-launching-synchronous-subtasks)

Can someone tell me the source code snippet for this `deadlock`?

**Comment by olii** [comment]
Suppose you have a single worker in your app. The worker reads a task from the queue and executes it. Now in the task, you have this code:

```shell
$ celery worker -A ... --concurrency 1
```

```python
@celery.task
def task():
    async_result = the_other_task.delay()
    # The following call blocks the worker until the result is available.
    value = async_result.get()
   print(value)

@celery.task
def the_other_task():
    return 'Done'
```

Execute the first task like this:
```
task.delay()
```


The worker sends the result to the broker and waits for its computation. While waiting, it cannot execute any other task. There are no other workers that can execute `the_other_task`. The worker would wait infinitely long.

**Reply by 954-Ivory** [reply]
Yep, but if I don't get `AsyncResult`, will it cause a deadlock?
```python3
@celery.task
def task():
    the_other_task.delay() # Just apply task here.

@celery.task
def the_other_task():
    return 'Done'
```

**Reply by auvipy** [reply]
you can try the python deadlock detector package
